src/parser/core/SharedCode.tsx

```diff
@@ -0,0 +1,231 @@
+   */
+  getExpiredProcs(buff: SpellInfo, spenderSpell: SpellInfo | SpellInfo[]) {
+    const events = this.getEvents(true, EventType.RemoveBuff, buff);
+
+    const filteredEvents = events.filter((e) => {
+      const castEvent = this.getEvents(
+        true,
+        EventType.Cast,
+        spenderSpell,
+        1,
+        e.timestamp + 1,
+        250,
+      )[0];
+      return !castEvent;
+    });
+    return filteredEvents;
+  }
+
+  /**
+   * @param buff the spell object for the proc's buff.
+   * @param spenderSpell the spell object (or an array of spell objects) that are used to spend the proc.
+   * @returns a numeric count of expired procs
+   */
+  countExpiredProcs(buff: SpellInfo, spenderSpell: SpellInfo | SpellInfo[]) {
+    return this.getExpiredProcs(buff, spenderSpell).length;
+  }
+
+  /**
+   * @param event the event that you want to check the target's health on.
+   * @returns the target's health percentage (in decimal, i.e. 0.75 = 75%)
+   */
+  getTargetHealth(event: AnyEvent) {
+    const target =
+      (event.type === EventType.Cast || event.type === EventType.BeginCast) && HasTarget(event)
+        ? encodeTargetString(event.targetID, event.targetInstance)
+        : undefined;
+    const damageEvents = this.getEvents(false, EventType.Damage, undefined, event.timestamp, 5000);
+    if (!damageEvents) {
+      return;
+    }
+
+    let relevantEvent;
+    if (target) {
+      relevantEvent = damageEvents.find(
+        (e) => HasTarget(e) && target === encodeTargetString(e.targetID, e.targetInstance),
+      );
+    } else {
+      relevantEvent = damageEvents[0];
+    }
+
+    if (relevantEvent && HasHitpoints(relevantEvent)) {
+      return relevantEvent.hitPoints / relevantEvent.maxHitPoints;
+    } else {
+      return undefined;
+    }
+  }
+
+  /**
+   * @param searchBackwards specify whether you want to search for events forwards or backwards from a particular timestamp (true for backwards, false for forwards. Default is backwards).
+   * @param eventType the event type to get (i.e. 'cast', 'begincast', EventType.Cast, EventType.BeginCast). Use EventType.Event for all events.
+   * @param spell the specific spell (or an array of spells) you are searching for. Leave undefined for all spells.
+   * @param count the number of events to get. Leave undefined for no limit.
+   * @param startTimestamp the timestamp to start searching from. Searches search backwards from the startTimestamp. Leave undefined for the end of the fight
+   * @param duration the amount of time in milliseconds to search. Leave undefined for no limit.
+   * @returns an array of events that meet the provided criteria
+   */
+  getEvents<ET extends EventType>(
+    searchBackwards: boolean = true,
+    eventType: ET,
+    spell?: SpellInfo | SpellInfo[],
+    count?: number,
+    startTimestamp: number = this.owner.fight.end_time,
+    duration?: number,
+    includePets: boolean = false,
+  ): Array<MappedEvent<ET>> {
```

Comment:
it is usually better to have an `options` object rather than a large number of optional/default parameters. positional parameters (especially ones that share a type but not a meaning, like `startTimestamp` and `duration`) are error prone in large quantities.

```suggestion
  type EventSearchOptions = {
    searchBackwards?: boolean;
    spell?: SpellInfo | SpellInfo[];
    count?: number;
    startTimestamp?: number;
    duration?: number;
    includePets?: boolean;
  }
  
  getEvents<ET extends EventType>(
    eventType: ET,
    { 
      searchBackwards, 
      spell, 
      count, 
      startTimestamp, 
      duration, 
      includePets
    }: EventSearchOptions = {}
  ): Array<MappedEvent<ET>> {
```

then callers use `getEvents(eventType, { spell: whatever })` and leave off the ones they don't care about. note that when taking this approach, you need to use `?` or `??` in the method body to default parameters, since providing default values in the params for this will be overwritten by the object.